packages/medusa/src/services/product-variant.ts

```diff
@@ -245,86 +254,197 @@ class ProductVariantService extends TransactionBaseService {
           )
         }
-      }
 
-      if (typeof metadata === "object") {
-        variant.metadata = setMetadata(variant as ProductVariant, metadata)
+        data = [{ variant: variant as ProductVariant, updateData: updateData }]
       }
 
-      if (typeof inventory_quantity === "number") {
-        variant.inventory_quantity = inventory_quantity as number
-      }
+      const result = await this.updateBatch(data)
+
+      return (Array.isArray(variantOrVariantIdOrData)
+        ? result
+        : result[0]) as unknown as TResult
+    })
+  }
+
+  protected async updateBatch(
+    variantData: {
+      variant: ProductVariant
+      updateData: UpdateProductVariantInput
+    }[]
+  ): Promise<ProductVariant[]> {
+    return await this.atomicPhase_(async (manager: EntityManager) => {
+      const variantRepo = manager.getCustomRepository(
+        this.productVariantRepository_
+      )
+
+      const variantPriceUpdateData = variantData
+        .filter((data) => isDefined(data.updateData.prices))
+        .map((data) => ({
+          variantId: data.variant.id,
+          prices: data.updateData.prices!,
+        }))
 
-      for (const [key, value] of Object.entries(rest)) {
-        variant[key] = value
+      if (variantPriceUpdateData.length) {
+        await this.updateVariantPrices(variantPriceUpdateData)
       }
 
-      const result = await variantRepo.save(variant)
+      const results: [ProductVariant, UpdateProductVariantInput, boolean][] =
+        await Promise.all(
+          variantData.map(async ({ variant, updateData }) => {
+            const { prices, options, ...rest } = updateData
+
+            const shouldUpdate = hasChanges(variant, rest)
+            const shouldEmitUpdateEvent =
+              shouldUpdate || !!options?.length || !!prices?.length
+
+            for (const option of options ?? []) {
+              await this.updateOptionValue(
+                variant.id!,
+                option.option_id,
+                option.value
+              )
+            }
+
+            const toUpdate: QueryDeepPartialEntity<ProductVariant> = {}
+
+            if (isObject(rest.metadata)) {
+              toUpdate["metadata"] = setMetadata(
+                variant as ProductVariant,
+                rest.metadata
+              ) as QueryDeepPartialEntity<Record<string, unknown>>
+              delete rest.metadata
+            }
+
+            if (Object.keys(rest).length) {
+              for (const [key, value] of Object.entries(rest)) {
+                if (variant[key] !== value) {
+                  toUpdate[key] = value
+                }
+              }
+            }
+
+            let result = variant
+
+            // No need to update if the nothing on the variant has been changed
+            if (shouldUpdate) {
+              const { id } = variant
+              const rawResult = await variantRepo.update({ id }, toUpdate)
+              result = variantRepo.create({
```

Comment:
question: why are we creating after we update?

Reply:
it create a new instance of an entity, it is not create in the sense of insert, it is more of a factory

Reply:
for reference:

```ts
create(entityClass, plainObjectOrObjects) {
        const metadata = this.connection.getMetadata(entityClass);
        if (!plainObjectOrObjects)
            return metadata.create(this.queryRunner);
        if (Array.isArray(plainObjectOrObjects))
            return plainObjectOrObjects.map((plainEntityLike) => this.create(entityClass, plainEntityLike));
        const mergeIntoEntity = metadata.create(this.queryRunner);
        this.plainObjectToEntityTransformer.transform(mergeIntoEntity, plainObjectOrObjects, metadata, true);
        return mergeIntoEntity;
    }
```